src/Microsoft.Health.Fhir.Shared.Api/Controllers/DataConvertController.cs

```diff
@@ -0,0 +1,173 @@
+using Microsoft.Health.Fhir.TemplateManagement.Models;
+using Microsoft.Health.Fhir.ValueSets;
+
+namespace Microsoft.Health.Fhir.Api.Controllers
+{
+    [ServiceFilter(typeof(AuditLoggingFilterAttribute))]
+    [ServiceFilter(typeof(OperationOutcomeExceptionFilterAttribute))]
+    [ServiceFilter(typeof(ValidateContentTypeFilterAttribute))]
+    [ValidateResourceTypeFilter]
+    [ValidateModelState]
+    public class DataConvertController : Controller
+    {
+        private readonly IMediator _mediator;
+        private readonly ILogger _logger;
+        private readonly DataConvertConfiguration _config;
+        private static HashSet<string> _supportedParams = GetSupportedParams();
+
+        private const char ImageRegistryDelimiter = '/';
+
+        public DataConvertController(
+            IMediator mediator,
+            IOptions<OperationsConfiguration> operationsConfig,
+            ILogger<DataConvertController> logger)
+        {
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+            EnsureArg.IsNotNull(operationsConfig?.Value?.DataConvert, nameof(operationsConfig));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _mediator = mediator;
+            _config = operationsConfig.Value.DataConvert;
+            _logger = logger;
+        }
+
+        [HttpPost]
+        [Route(KnownRoutes.DataConvert)]
+        [AuditEventType(AuditEventSubType.DataConvert)]
+        public async Task<IActionResult> DataConvert([FromBody] Parameters inputParams)
+        {
+            CheckIfDataConvertIsEnabled();
+
+            ValidateParams(inputParams);
+
+            string inputData = ReadStringParameter(inputParams, DataConvertProperties.InputData);
+            string templateCollectionReference = ReadStringParameter(inputParams, DataConvertProperties.TemplateCollectionReference);
+            string entryPointTemplate = ReadStringParameter(inputParams, DataConvertProperties.EntryPointTemplate);
+            DataConvertInputDataType inputDataType = ReadEnumParameter<DataConvertInputDataType>(inputParams, DataConvertProperties.InputDataType);
+
+            if (!ImageInfo.IsValidImageReference(templateCollectionReference))
+            {
+                _logger.LogInformation("Templates collection reference format is invalid.");
+                throw new RequestNotValidException(string.Format(Resources.InvalidTemplateCollectionReference, templateCollectionReference));
+            }
+
+            string registryServer = ExtractRegistryServer(templateCollectionReference);
+            var dataConvertRequest = new DataConvertRequest(inputData, inputDataType, registryServer, templateCollectionReference, entryPointTemplate);
+            DataConvertResponse response = await _mediator.Send(dataConvertRequest, cancellationToken: default);
+
+            return new ContentResult
+            {
+                Content = response.Resource,
+                ContentType = "application/json",
```

Comment:
> ContentType = "application/json", [](start = 15, length = 34)

Should data conversion support [general parameters](https://www.hl7.org/fhir/http.html#parameters) like `_format`?
What would happen if I specify in header `Accept` = `application/xml`,  would it convert it via OutputFormatter?

Reply:
No, for current templates, we will only return json format (output is defined by template files). Do you have any suggestions here?

Reply:
Right, so the response in not guarateed to be valid FHIR, or even valid JSON, so we might not be able to parse it.

Reply:
If customers provide a normal template in a way like the default one, the output is a valid JSON, yet not ensured to be a valid FHIR (there still minor gaps in mapping hl7v2 to FHIR), and this can be improved by refining templates in the future.
However, if customer provide an invalid template collection, the result could be invalid JSON or any content generated from his templates. This is somewhat similar to the case that user provided an invalid FHIR input?

Reply:
I imagine using $data-convert to convert FHIR to HL7v2, or non-JSON formats in future. Co, can we loosen the constraint here?

Reply:
Sure. Let's return raw content here.